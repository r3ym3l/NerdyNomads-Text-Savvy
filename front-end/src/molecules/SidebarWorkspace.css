.SidebarWorkspace-header{
    width: calc(100% - 24px);
    padding: 10px 12px;

    display: flex;
    flex-direction: row;

    justify-content: space-between;

    background-color: rgba(65, 148, 98, 0.25);
    font-weight: 700;
    font-size: 18px;
}

.SidebarWorkspace {
    overflow: auto;
}

.SidebarWorkspace-add-icon{
    user-select: none;
    cursor: pointer;
}

.SidebarWorkspace-add{
    width: 100%;

    display: flex;
    flex-direction: row;

    justify-content: center;

    background-color: #e7e7e7;
    font-weight: 700;
    font-size: 18px;
}

.SidebarWorkspace-add-input{
    border-style: hidden;
    outline: none;
    resize: none; 
    font-family: inherit;
    font-size: 16px;
    border-radius: 4px;
    width: 100%;
    padding: 5px;
    margin: 10px;
}