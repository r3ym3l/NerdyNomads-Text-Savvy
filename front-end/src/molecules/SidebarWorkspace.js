import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { useAuth0 } from "@auth0/auth0-react";

import { AddWorkspaceIcon } from "../atoms/icons";
import SidebarWorkspaceItem from "../atoms/SidebarWorkspaceItem";
import WorkspaceSettings from "../organisms/WorkspaceSettings";
import { addNewWorkspace, getCollabWorkspaces, getOwnedWorkspaces, updateAccountWorkspaces } from "../util/requests";
import "./SidebarWorkspace.css";


const componentName = "SidebarWorkspace";
function SidebarWorkspace( {onSelectWorkspace, accountId} ) {  
  
  const [ showAddWorkspace, setShowAddWorkspace ] = useState(false);
  const [ workspaceList, setWorkspaceList ] = useState([]);
  const [ showWorkspaceSettingPopup, setShowWorkspaceSettingPopup ] = useState(false);

  const { user } = useAuth0();
  const [ editingWorkspaceId, setEditingWorkspaceId ] = useState();
  const [ currentWorkspaceId, setCurrentWorkspaceId ] = useState("");

  // Updates which workspace is highlighted and selected in the dashboard
  const updateSelectedWorkspace = workspaceId =>{
    setCurrentWorkspaceId(workspaceId);
    setEditingWorkspaceId(workspaceId);
    onSelectWorkspace(workspaceId);
    localStorage.setItem("currentWorkspaceId", workspaceId);
  };

  const handleWorkspaceSubmit = async (e) => {
    if (e.key === "Enter") {
      e.preventDefault();

      const response = await addNewWorkspace(e.target.value, accountId);
      const newWorkspaceList = [response.data, ...workspaceList];
      setWorkspaceList(newWorkspaceList);

      let workspaceIds = newWorkspaceList.map(workspace => workspace._id);
      updateAccountWorkspaces(accountId, workspaceIds);
      window.location.reload(false);  // to update extension with new workspace
      e.target.value = "";
      setShowAddWorkspace(false);
      updateSelectedWorkspace(response.data._id);
    }
  };

  const addWorkspaceInput = showAddWorkspace ?
    <div className={`${componentName}-add`}>
      <input 
        type="text" 
        autoFocus 
        placeholder="Enter Workspace Name"
        className={`${componentName}-add-input`}
        onKeyPress={(e) => handleWorkspaceSubmit(e)}/>
    </div> 
    : <></>;

  const handleAddWorkspace = () => setShowAddWorkspace(prevShow => !prevShow);

  const handleOnWorkspaceEdit = workspaceId => {
    setShowWorkspaceSettingPopup(true);
    updateSelectedWorkspace(workspaceId);
  };

  const handleOnChangeVisibility = (visible) => setShowWorkspaceSettingPopup(visible);

  const renderList = () => (
    workspaceList && workspaceList.map( ({_id, name}) => 
      <SidebarWorkspaceItem key={_id} id={_id} selected={_id === currentWorkspaceId} name={name} onEdit={() => handleOnWorkspaceEdit(_id)} onClickWorkspace={updateSelectedWorkspace}/>)
  );

  useEffect( async () => {
    if (accountId) {
      let ownedWorkspaces = await getOwnedWorkspaces(accountId);
      let collabWorkspaces = await getCollabWorkspaces(user.email);
      let workspaces = (ownedWorkspaces.data).concat(collabWorkspaces.data);
    
      setWorkspaceList(workspaces);
    }
  }, [accountId]); 

  useEffect(() => {
    if (workspaceList.length > 0){

      var retrievedWorkspace = localStorage.getItem("currentWorkspaceId");
      if (retrievedWorkspace == null) {
        localStorage.setItem("currentWorkspaceId" , workspaceList[0]._id);
      }

      retrievedWorkspace = localStorage.getItem("currentWorkspaceId");
      onSelectWorkspace(retrievedWorkspace);
      setCurrentWorkspaceId(retrievedWorkspace);
    } // else, do not show any workspace because there aren't any workspaces
  }, [workspaceList]);

  useEffect(() => {
    if (currentWorkspaceId != "") {
      onSelectWorkspace(currentWorkspaceId);
      setEditingWorkspaceId(currentWorkspaceId);
      localStorage.setItem("currentWorkspaceId", currentWorkspaceId);
    }
  }, [currentWorkspaceId]);

  const header = <div className={`${componentName}-header`}>
    <span>My Workspaces</span>
    <AddWorkspaceIcon className={`${componentName}-add-icon`} onClick={handleAddWorkspace}/>
  </div>;

  return (
    <div className={`${componentName}`}>
      {header}
      {addWorkspaceInput}
      {renderList()}
      {/* Workspace Settings Pop Up */}
      { showWorkspaceSettingPopup && <WorkspaceSettings workspaceId={editingWorkspaceId} onChangeVisibility={(visible) => handleOnChangeVisibility(visible)}/>}
    </div>
  );
}

SidebarWorkspace.propTypes = {
  onSelectWorkspace: PropTypes.func.isRequired,
  accountId: PropTypes.string.isRequired
};


export default SidebarWorkspace;