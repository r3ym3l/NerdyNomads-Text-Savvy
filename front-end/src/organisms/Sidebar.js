/* global chrome */
import React from "react";
import PropTypes from "prop-types";

import { useAuth0 } from "@auth0/auth0-react";

import { LogoutIcon, LogoIcon } from "../atoms/icons";
import SidebarWorkspace from "../molecules/SidebarWorkspace";

import "./Sidebar.css";

function Sidebar({ onClickWorkspace, accountId }) {
  const { logout } = useAuth0();

  const spacing = <div className="SideBar-spacing"/>;
  const divider = <div className="SideBar-divider"/>;
  const dividerLight = <div className="SideBar-divider-light"/>;
  const componentName = "SideBar";

  const handleGoToWorkspace = (id) => onClickWorkspace(id);
  
  const handleLogout = () => {
    if (chrome.runtime !== undefined) {
      // clear auth0Id and send to web extension
      window.localStorage.setItem("auth0Id", "");
      chrome.runtime.sendMessage(`${process.env.REACT_APP_EXTENSION_ID}`, { messageFromWeb: window.localStorage });
    }
    logout();
  };

  return (
    <div className={`${componentName}`}>
      <div className={`${componentName}-header`}>
        <LogoIcon className={`${componentName}-logo`}/>
        <div className={`${componentName}-logo-text`}>TextSavvy</div>
      </div>
      {divider}
      <SidebarWorkspace onSelectWorkspace={handleGoToWorkspace} accountId={accountId}/>
      {spacing}
      {dividerLight}
      <div className={`${componentName}-logout ${componentName}-option`} onClick={handleLogout}>
        <div className={`${componentName}-option-logo`}><LogoutIcon/></div>
        <div className={`${componentName}-option-text`}>Logout</div>
      </div>
    </div>
  );
}

Sidebar.propTypes = {
  onClickWorkspace: PropTypes.func.isRequired,
  accountId: PropTypes.string.isRequired
};


export default Sidebar;