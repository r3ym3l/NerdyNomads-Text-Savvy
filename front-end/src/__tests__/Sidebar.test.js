/* eslint-disable no-undef */
/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/display-name */
import React from "react";
import { shallow } from "enzyme";

import Sidebar from "../organisms/Sidebar";
import { useAuth0 } from "@auth0/auth0-react";
import SidebarWorkspace from "../molecules/SidebarWorkspace";

jest.mock("@auth0/auth0-react"); 
jest.mock("../molecules/SidebarWorkspace", () => {
  return () => <div />;
});

beforeEach(() => {
  useAuth0.mockReturnValue({
    logout: jest.fn(),
    user: {
      nickname: "Test"
    } 
  });
});

test("Should match the snapshot.", () => { 
  const mockFunc = () => null;
  let wrapper = shallow(<Sidebar onClickWorkspace={mockFunc} accountId=""/>);
  expect(wrapper.html()).toMatchSnapshot();
});

test("Clicking the logout button should call Auth0's logout function.", () => {
  global.chrome = {
    runtime: {
      sendMessage: jest.fn()
    }
  };
  const mockFunc = () => null;
  let wrapper = shallow(<Sidebar onClickWorkspace={mockFunc} accountId=""/>);

  let logoutDiv = wrapper.find(".SideBar-logout");
  expect(logoutDiv.length).toEqual(1);
  logoutDiv.simulate("click");
  expect(useAuth0().logout).toHaveBeenCalledTimes(1);
});

test("Should display the logout button.", () => {
  const mockFunc = () => null;
  let wrapper = shallow(<Sidebar onClickWorkspace={mockFunc} accountId=""/>);

  let logoutDiv = wrapper.find(".SideBar-logout");
  expect(logoutDiv.length).toEqual(1);

  let logoutText = wrapper.find(".SideBar-option-text").at(0);
  expect(logoutText.length).toEqual(1);
  expect(logoutText.text()).toEqual("Logout");
});

test("Should display the SidebarWorkspace component.", () => {
  const mockFunc = () => null;
  let wrapper = shallow(<Sidebar onClickWorkspace={mockFunc} accountId=""/>);
  
  expect(wrapper.containsMatchingElement(<SidebarWorkspace />)).toEqual(true);
});
