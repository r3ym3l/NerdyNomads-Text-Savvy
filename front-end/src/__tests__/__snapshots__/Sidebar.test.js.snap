// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match the snapshot. 1`] = `"<div class=\\"SideBar\\"><div class=\\"SideBar-header\\"><img alt=\\"text savvy logo\\" src=\\"/assets/images/logo.svg\\" class=\\"SideBar-logo\\"/><div class=\\"SideBar-logo-text\\">TextSavvy</div></div><div class=\\"SideBar-divider\\"></div><div></div><div class=\\"SideBar-spacing\\"></div><div class=\\"SideBar-divider-light\\"></div><div class=\\"SideBar-logout SideBar-option\\"><div class=\\"SideBar-option-logo\\"><img alt=\\"logout\\" src=\\"/assets/images/logout.svg\\" class=\\"icon\\"/></div><div class=\\"SideBar-option-text\\">Logout</div></div></div>"`;
